google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java

```diff
@@ -1168,11 +1173,82 @@ public Boolean call() {
-      throws InterruptedException, JobException {
+      throws InterruptedException {
     Job.checkNotDryRun(configuration, "query");
+
+    // If all parameters passed in configuration are supported by the query() method on the backend,
+    // put on fast path
+    QueryRequestInfo requestInfo = new QueryRequestInfo(configuration);
+    if (requestInfo.isFastQuerySupported()) {
+      String projectId = getOptions().getProjectId();
+      QueryRequest content = requestInfo.toPb();
+      return fastQuery(projectId, content, options);
+    }
+    // Otherwise, fall back to the existing create query job logic
     return create(JobInfo.of(configuration), options).getQueryResults();
   }
 
+  private TableResult fastQuery(
+      final String projectId, final QueryRequest content, JobOption... options)
+      throws InterruptedException {
+    com.google.api.services.bigquery.model.QueryResponse results;
+    try {
+      results =
+          runWithRetries(
+              new Callable<com.google.api.services.bigquery.model.QueryResponse>() {
+                @Override
+                public com.google.api.services.bigquery.model.QueryResponse call() {
+                  return bigQueryRpc.fastQuery(projectId, content);
+                }
+              },
+              getOptions().getRetrySettings(),
+              EXCEPTION_HANDLER,
+              getOptions().getClock());
+    } catch (RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+
+    // If fast query completed and has only one page in results
+    if (results.getJobComplete() && results.getPageToken() == null) {
+      // If there are errors, BigQueryException is thrown
+      ImmutableList.Builder<BigQueryError> errors = ImmutableList.builder();
+      if (results.getErrors() != null) {
+        List<BigQueryError> bigQueryErrors =
+            Lists.transform(results.getErrors(), BigQueryError.FROM_PB_FUNCTION);
+        // Throwing BigQueryException since there may be no JobId and we want to stay consistent
+        // with the case where there there is a HTTP error
+        throw new BigQueryException(bigQueryErrors);
+      }
+
+      // If there is no error, we construct TableResult
+      TableSchema schemaPb = results.getSchema();
+
+      Long numRows;
+      if (results.getNumDmlAffectedRows() == null && results.getTotalRows() == null) {
+        numRows = 0L;
+      } else if (results.getNumDmlAffectedRows() != null) {
+        numRows = results.getNumDmlAffectedRows();
+      } else {
+        numRows = results.getTotalRows().longValue();
+      }
+
+      return new TableResult(
+          schemaPb == null ? null : Schema.fromPb(schemaPb),
+          numRows,
+          new PageImpl<>(
+              new TableDataPageFetcher(null, getOptions(), null, optionMap(options)),
+              null,
+              transformTableData(results.getRows())));
+    } else {
```

Comment:
Flip the if condition to move this branch up and since you have a return in this branch you can remove the nesting.

if (...) {
line 1243 - line 1247;
return result;
}

line 1213 - line 1235

Reply:
flipped but not sure about eliminating nesting.